realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java

```diff
@@ -226,56 +223,273 @@ protected RemoteInsertManyResult mapSuccess(Object result) {
+            case REPLACE:
+                if (options == null) {
+                    nativeFindOneAndReplace(nativePtr, encodedFilter, encodedUpdate, callback);
+                } else {
+                    nativeFindOneAndReplaceWithOptions(nativePtr, encodedFilter, encodedUpdate, encodedProjection, encodedSort, options.isUpsert(), options.isReturnNewDocument(), callback);
+                }
+                break;
+            case DELETE:
+                if (options == null) {
+                    nativeFindOneAndDelete(nativePtr, encodedFilter, callback);
+                } else {
+                    nativeFindOneAndDeleteWithOptions(nativePtr, encodedFilter, encodedProjection, encodedSort, options.isUpsert(), options.isReturnNewDocument(), callback);
+                }
+                break;
+        }
+
         return ResultHandler.handleResult(success, error);
     }
 
-    public RemoteDeleteResult deleteMany(final Bson filter) {
-        AtomicReference<RemoteDeleteResult> success = new AtomicReference<>(null);
+    private DeleteResult deleteInternal(final DeleteType type, final Bson filter) {
+        AtomicReference<DeleteResult> success = new AtomicReference<>(null);
         AtomicReference<ObjectServerError> error = new AtomicReference<>(null);
-        OsJNIResultCallback<RemoteDeleteResult> callback = new OsJNIResultCallback<RemoteDeleteResult>(success, error) {
+        OsJNIResultCallback<DeleteResult> callback = new OsJNIResultCallback<DeleteResult>(success, error) {
             @Override
-            protected RemoteDeleteResult mapSuccess(Object result) {
-                return new RemoteDeleteResult((Long) result);
+            protected DeleteResult mapSuccess(Object result) {
+                return new DeleteResult((Long) result);
             }
         };
 
         String jsonDocument = JniBsonProtocol.encode(filter, codecRegistry);
-        nativeDeleteMany(nativePtr, jsonDocument, callback);
+        switch (type) {
+            case ONE:
+                nativeDeleteOne(nativePtr, jsonDocument, callback);
+                break;
+            case MANY:
+                nativeDeleteMany(nativePtr, jsonDocument, callback);
+                break;
+        }
         return ResultHandler.handleResult(success, error);
     }
 
-    private <T> T findOneSuccessMapper(@Nullable Object result, Class<T> resultClass) {
+    private <T> T findSuccessMapper(@Nullable Object result, Class<T> resultClass) {
         if (result == null) {
             return null;
         } else {
             return JniBsonProtocol.decode((String) result, resultClass, codecRegistry);
         }
     }
 
+    private enum UpdateType {
+        ONE, MANY
+    }
+
+    private enum DeleteType {
+        ONE, MANY
+    }
+
+    private enum FindOneAndType {
+        UPDATE, REPLACE, DELETE
+    }
+
     private static native long nativeGetFinalizerMethodPtr();
     private static native void nativeCount(long remoteMongoCollectionPtr,
```

Comment:
See previous comment about maybe combining these into methods taking a selector. You already have the differentiation modelled in the `FindOneAndType` enums, etc.

Reply:
I will address this in the next PR.